dnsdbq.c

```diff
@@ -1379,18 +1383,101 @@ query_launcher(qdesc_ct qdp, qparam_ct qpp, writer_t writer) {
 			fence.first_before = qpp->before;
 		}
 	}
-	launch(query, &fence);
+
+	/* branch on rrtype. */
+	if (qdp->rrtype == NULL) {
+		/* no rrtype string given, let makepath set it to "any". */
+		char *path = makepath(qdp);
+		launch_fetch(query, path, &fence);
+		DESTROY(path);
+	} else if ((msg = multitype_coherency(qdp->rrtype)) != NULL) {
+		fprintf(stderr, "%s: multitype_coherency failed: %s\n",
+			program_name, msg);
+		DESTROY(query);
+		return NULL;
+	} else {
+		/* rrtype string was given, parse comma separated list. */
+		char *rrtypes = strdup(qdp->rrtype);
+		char *saveptr = NULL;
+		int nfetches = 0;
+		for (char *rrtype = strtok_r(rrtypes, ",", &saveptr);
+		     rrtype != NULL;
+		     rrtype = strtok_r(NULL, ",", &saveptr))
+		{
+			struct qdesc qd = {
+				.mode = qdp->mode,
+				.thing = qdp->thing,
+				.rrtype = rrtype,
+				.bailiwick = qdp->bailiwick,
+				.pfxlen = qdp->pfxlen
+			};
+			char *path = makepath(&qd);
+			launch_fetch(query, path, &fence);
+			nfetches++;
+			DESTROY(path);
+		}
+		if (nfetches > 1)
+			query->multitype = true;
+		DESTROY(rrtypes);
+	}
+
+	/* finish query initialization, link it up, and return it. */
+	query->writer = writer;
+	writer = NULL;
+	query->params = *qpp;
+	query->next = query->writer->queries;
+	query->writer->queries = query;
+	query->descrip = makepath(qdp);
+	query->mode = qdp->mode;
 	return query;
 }
 
-/* launch -- actually launch a query job, given a command and time fences.
+/* multitype_coherency -- return an error text if these rrtypes don't mix.
+ */
+static const char *
+multitype_coherency(const char *input) {
+	char *rrtypes = strdup(input);
+	char *saveptr = NULL;
+	bool some = false, any = false,
+		some_dnssec = false, any_dnssec = false;
+	for (char *rrtype = strtok_r(rrtypes, ",", &saveptr);
+	     rrtype != NULL;
+	     rrtype = strtok_r(NULL, ",", &saveptr))
+	{
+		for (char *p = rrtype; *p != '\0'; p++)
+			if (isupper(*p))
+				*p = (char) tolower(*p);
+		if (strcmp(rrtype, "any") == 0)
+			any = true;
+		else if (strcmp(rrtype, "any-dnssec") == 0)
+			any_dnssec = true;
+		else if (strcmp(rrtype, "ds") == 0 ||
```

Comment:
Should also consider CDS, CDNSKEY, and TA as DNSSEC.
Should also check for TYPE### equivalents.